### PR TITLE
Add docstrings for LabelModel, Baseline, Logger

### PR DESCRIPTION
**Pull request by paroma**
* Add docstrings for `labeling/model`

* Switch LabelModel loss functions to be private

__Test Plan__
Edit test to match private loss function

### CODE REVIEW

File: snorkel/labeling/model/label_model.py [hunk 1]
```diff
@@ -63,17 +107,19 @@ def _check_L(self, L: Matrix) -> np.ndarray:
...
 
-        Returns:
-            L_ind: An [n,m*k] dense np.ndarray with values in {0,1}
+        Parameters
+        ----------
+        L
+            An [n,m] label matrix with values in {0,1,...,k}
 
-        Note that no column is required for 0 (abstain) labels.
+        Returns
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: Nit: add space

File: snorkel/labeling/model/label_model.py [hunk 2]
```diff
@@ -84,12 +130,23 @@ def _create_L_ind(self, L: Matrix) -> np.ndarray:
...
         for whether a certain source or clique of sources voted in a certain
         pattern.
 
-        Args:
-            L: An [n,m] scipy.sparse label matrix with values in {0,1,...,k}
+        Parameters
+        ----------
+        L
+            An [n,m] label matrix with values in {0,1,...,k}
+        higher_order: optional
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: remove optional

File: snorkel/labeling/model/label_model.py [hunk 3]
```diff
@@ -132,7 +189,13 @@ def _get_augmented_label_matrix(
             return L_ind
 
     def _build_mask(self) -> None:
-        """Build mask applied to O^{-1}, O for the matrix approx constraint"""
+        """Build mask applied to O^{-1}, O for the matrix approx constraint.
+
+        Attributes
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: No attributes for function

File: snorkel/labeling/model/label_model.py [hunk 4]
```diff
@@ -147,18 +210,21 @@ def _build_mask(self) -> None:
...
-
-        Note that we only include the k non-abstain values of each source,
-        otherwise the model not minimal --> leads to singular matrix
+        """Generate overlaps and conflicts matrix from label matrix.
+
+        Parameters
+        ----------
+        L
+            An [n,m] label matrix with values in {0,1,...,k}
+        higher_order: optional
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: rm optional

File: snorkel/labeling/model/label_model.py [hunk 5]
```diff
@@ -147,18 +210,21 @@ def _build_mask(self) -> None:
...
+            Whether to include higher-order correlations (e.g. LF pairs) in matrix, by default False
         """
         L_aug = self._get_augmented_label_matrix(L, higher_order=higher_order)
         self.d = L_aug.shape[1]
         self.O = torch.from_numpy(L_aug.T @ L_aug / self.n).float()
 
     def _init_params(self) -> None:
-        """Initialize the learned params
+        r"""Initialize the learned params.
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: regex?

File: snorkel/labeling/model/label_model.py [hunk 6]
```diff
@@ -199,9 +270,10 @@ def _init_params(self) -> None:
         # Build the mask over O^{-1}
         self._build_mask()
 
-    def get_conditional_probs(self, source: Optional[int] = None) -> np.ndarray:
-        """Returns the full conditional probabilities table as a numpy array,
-        where row i*(k+1) + ly is the conditional probabilities of source i
+    def _get_conditional_probs(self, source: Optional[int] = None) -> np.ndarray:
+        r"""Return the full conditional probabilities table.
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: same

File: snorkel/labeling/model/label_model.py [hunk 7]
```diff
@@ -211,6 +283,16 @@ def get_conditional_probs(self, source: Optional[int] = None) -> np.ndarray:
         probability and adding in to mu.
 
         If `source` is not None, returns only the corresponding block.
+
+        Parameters
+        ----------
+        source : optional
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: rm optional

File: snorkel/labeling/model/label_model.py [hunk 8]
```diff
@@ -237,11 +319,26 @@ def get_conditional_probs(self, source: Optional[int] = None) -> np.ndarray:
             return c_probs
 
     def get_accuracies(self, probs: Optional[np.ndarray] = None) -> np.ndarray:
-        """Returns the vector of LF accuracies, computed using get_conditional_probs"""
+        """Return the vector of LF accuracies.
+
+        Example
+        -------
+        label_model.get_accuracies()
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: not very useful

File: snorkel/labeling/model/label_model.py [hunk 9]
```diff
@@ -250,10 +347,22 @@ def get_accuracies(self, probs: Optional[np.ndarray] = None) -> np.ndarray:
...
     def predict_proba(self, L: Matrix) -> np.ndarray:
-        """Returns the [n,k] matrix of label probabilities P(Y | \lambda)
-
-        Args:
-            L: An [n,m] scipy.sparse label matrix with values in {0,1,...,k}
+        r"""Return label probabilities P(Y | \lambda).
+
+        Examples
+        --------
+        label_model.train_model(L)
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: not very useful

File: snorkel/labeling/model/label_model.py [hunk 10]
```diff
@@ -270,16 +379,23 @@ def predict_proba(self, L: Matrix) -> np.ndarray:
     # (for better or worse). The unused *args make these compatible with the
     # Classifer._train() method which expect loss functions to accept an input.
 
-    def loss_l2(self, l2: float = 0) -> torch.Tensor:
-        """L2 loss centered around mu_init, scaled optionally per-source.
+    def _loss_l2(self, l2: float = 0) -> torch.Tensor:
+        r"""L2 loss centered around mu_init, scaled optionally per-source.
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: regex

File: snorkel/labeling/model/label_model.py [hunk 11]
```diff
@@ -270,16 +379,23 @@ def predict_proba(self, L: Matrix) -> np.ndarray:
...
         In other words, diagonal Tikhonov regularization,
             ||D(\mu-\mu_{init})||_2^2
         where D is diagonal.
 
-        Args:
-            - l2: A float or np.array representing the per-source regularization
-                strengths to use
+        Parameters
+        ----------
+        l2 : optional
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: rm optional

File: snorkel/labeling/model/label_model.py [hunk 12]
```diff
@@ -289,15 +405,28 @@ def loss_l2(self, l2: float = 0) -> torch.Tensor:
         # Note that mu is a matrix and this is the *Frobenius norm*
         return torch.norm(D @ (self.mu - self.mu_init)) ** 2
 
-    def loss_mu(self, l2: float = 0) -> torch.Tensor:
+    def _loss_mu(self, l2: float = 0) -> torch.Tensor:
+        r"""Overall mu loss.
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: regex

File: snorkel/labeling/model/label_model.py [hunk 13]
```diff
@@ -289,15 +405,28 @@ def loss_l2(self, l2: float = 0) -> torch.Tensor:
...
+
+        Parameters
+        ----------
+        l2 : optional
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: rm optional

File: snorkel/labeling/model/label_model.py [hunk 14]
```diff
@@ -289,15 +405,28 @@ def loss_l2(self, l2: float = 0) -> torch.Tensor:
...
+            A float or np.array representing the per-source regularization
+                strengths to use, by default 0
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: no tab on second line. i think pydocstyle should catch this

File: snorkel/labeling/model/label_model.py [hunk 15]
```diff
@@ -400,20 +529,34 @@ def train_model(
...
-        No dependencies (conditionally independent sources): Estimate mu
-        subject to constraints:
-            (1a) O_{B(i,j)} - (mu P mu.T)_{B(i,j)} = 0, for i != j, where B(i,j)
-                is the block of entries corresponding to sources i,j
-            (1b) np.sum( mu P, 1 ) = diag(O)
+        """Train label model.
+
+        Train label model to estimate mu, the parameters related to accuracies of LFs.
+
+        Examples
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: Same

File: snorkel/labeling/model/label_model.py [hunk 16]
```diff
@@ -400,20 +529,34 @@ def train_model(
...
+
+        label_model.train(L)
+        label_model.train(L, Y_dev=Y_dev)
+        label_model.train(L, class_balance=[0.7, 0.3])
+
+        Parameters
+        ----------
+        L_train
+            An [n,m] matrix with values in {0,1,...,k}
+        Y_dev : optional
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: rm opt

File: snorkel/labeling/model/label_model.py [hunk 17]
```diff
@@ -489,10 +632,44 @@ def train_model(
             print("Finished Training")
 
     def save(self, destination: str, **kwargs: Any) -> None:
+        """Save label model.
+
+        Examples
+        --------
+        label_model.save('./saved_label_model')
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: not very useful

File: snorkel/labeling/model/label_model.py [hunk 18]
```diff
@@ -489,10 +632,44 @@ def train_model(
...
+
+        Parameters
+        ----------
+        destination
+            File location for saving model
+
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: extra space

File: snorkel/labeling/model/label_model.py [hunk 19]
```diff
@@ -489,10 +632,44 @@ def train_model(
...
         with open(destination, "wb") as f:
             torch.save(self, f, **kwargs)
 
     @staticmethod
     def load(source: str, **kwargs: Any) -> Any:
+        """Load existing label model.
+
+        Examples
+        --------
+        label_model = LabelModel()
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: not very useful

File: snorkel/labeling/model/label_model.py [hunk 20]
```diff
@@ -489,10 +632,44 @@ def train_model(
...
+        label_model.load('./saved_label_model')
+
+        Parameters
+        ----------
+        source
+            File location from where to load model
+
```
Comment (henryre) [inline snorkel/labeling/model/label_model.py]: extra space

File: snorkel/labeling/model/logger.py [hunk 21]
```diff
@@ -3,22 +3,66 @@
...
+    Parameters
+    ----------
+    log_train_every
+        Number of units at which to log model
+
+    Attributes
+    ----------
+    log_train_every
+        Number of units at which to log model
+
```
Comment (henryre) [inline snorkel/labeling/model/logger.py]: extra space

File: snorkel/labeling/model/logger.py [hunk 22]
```diff
@@ -3,22 +3,66 @@
 
 
 class Logger:
+    """Class for logging LabelModel.
+
+    Parameters
+    ----------
+    log_train_every
+        Number of units at which to log model
+
```
Comment (henryre) [inline snorkel/labeling/model/logger.py]: Parameters?

File: snorkel/labeling/model/logger.py [hunk 23]
```diff
@@ -3,22 +3,66 @@
...
     def log(self, metrics_dict: Dict[str, float]) -> None:
-        """Print calculated metrics and optionally write to file (json/tb)"""
+        """Print calculated metrics and optionally write to file (json/tb).
+
+        Parameters
+        ----------
+        metrics_dict
+            Dictionary of metric names (keys) and values to log
+
+        Example
```
Comment (henryre) [inline snorkel/labeling/model/logger.py]: Same

File: test/labeling/model/test_label_model.py [hunk 24]
```diff
@@ -227,26 +227,26 @@ def test_loss(self):
 
         # l2_loss = l2*M*K*||mu - mu_init||_2 = 3*2*(0.05^2) = 0.03
         self.assertAlmostEqual(
-            label_model.loss_l2(l2=1.0).detach().numpy().ravel()[0], 0.03
+            label_model._loss_l2(l2=1.0).detach().numpy().ravel()[0], 0.03
```
Comment (henryre) [inline test/labeling/model/test_label_model.py]: Nit: `label_model._loss_l2(l2=1.0).item()`

use `.item()` for all below changes

**Comment by henryre** [comment]
Great job with the examples for the baselines!! Quite a few formatting issues here that I think pydocstyle is supposed to catch. Was running it through tox not showing a warning?

File: snorkel/labeling/model/label_model.py (hunk 5 above)
Comment (paroma) [reply snorkel/labeling/model/label_model.py]: pydocstyle complains if there's `\` in the docstring, should I wrap math in something to prevent the error?

Comment (henryre) [reply snorkel/labeling/model/label_model.py]: which error code is it?
